### ISSUE TITLE
@emotion/react: Support using theme props with `css` API

### ISSUE THREAD

**Issue by hantatsang** [emotion-js/emotion#2252]
**The problem**

Existing `css` from `@emotion/react` doesn't support accessing theme props as such:

```js
const mixin = css`
  color: ${props => props.theme.colors.primary}
`'
```

**Proposed solution**

Please support it as it is helpful for React developers:

- Allow use cases in theme definitions, mixins, helpers, etc. where accessing theme props is required
- Users migrating to Emotion from Styled Components will be blocked by the lack of this feature

No specific solution from me but `styled-components` is a good example.

**Alternative solutions**

N/A

**Additional context**

Similar issue: #801

**Comment by ivancuric** [comment]
I see a few other options:
- The[ `useTheme` ](https://emotion.sh/docs/theming#usetheme) can be used inside a component.
- Just import the theme object and use it raw. The only difference (as far as I see) is the lack of reactivity, which can be solved by hooking the component up to the `useTheme` hook.

**Comment by Andarist** [comment]
You can also just refactor this to:
```js
const mixin = theme => css`
  color: ${theme.colors.primary}
`'
```

**Comment by rickstaa** [comment]
@Andarist Is there a typescript equivalent to the solution you are providing. I can get it to work in javascript or by using the [useTheme](https://emotion.sh/docs/theming#usetheme) hook while forwarding my theme to the css function. But I did not yet find a way to use your solution in typescript. I tried the following syntax:

```js

import { Theme } from "@material-ui/core/styles";

const mixin = (theme: Theme) => css`
  color: ${theme.colors.primary}
`'
```
But this gives me the following error:

```ts
No overload matches this call.
  Overload 1 of 2, '(template: TemplateStringsArray, ...args: CSSInterpolation[]): SerializedStyles', gave the following error.
    The argument of type 'Color' is not assignable to parameter of type 'CSSInterpolation'.
      Type 'Color' is not assignable to type 'CSSObject'.
        Index signature is missing in type 'Color'.
  Overload 2 of 2, '(...args: CSSInterpolation[]): SerializedStyles', gave the following error.
    Argument of type 'TemplateStringsArray' is not assignable to parameter of type 'CSSInterpolation'.
      Type 'TemplateStringsArray' is not assignable to type 'CSSObject'.
        Types of property 'filter' are incompatible.
          Type '{ <S extends string>(predicate: (value: string, index: number, array: readonly string[]) => value is S, thisArg?: any): S[]; (predicate: (value: string, index: number, array: readonly string[]) => unknown, thisArg?: any): string[]; }' is not assignable to type 'Filter | (Filter | undefined)[] | Filter[] | undefined'.
            Type '{ <S extends string>(predicate: (value: string, index: number, array: readonly string[]) => value is S, thisArg?: any): S[]; (predicate: (value: string, index: number, array: readonly string[]) => unknown, thisArg?: any): string[]; }' is missing the following properties from type 'Filter[]': pop, push, concat, join, and 27 more.  TS2769
```

I also tried to use the ThemeProps, but this interface explained in #1670 doesn't seem to be present inside v11. I also tried creating the ThemeProps interface myself using:

```ts
interface ThemeProps {
  theme: Theme;
  [props: string]: any;
}
```

But this gives me the following error:

```bash
No overload matches this call.
  Overload 1 of 2, '(props: { component: ElementType<any>; } & Omit<AnchorHTMLAttributes<HTMLAnchorElement>, "color"> & Omit<TypographyProps<"span", {}>, "color" | ... 2 more ... | "variant"> & { ...; } & CommonProps & Omit<...>): Element', gave the following error.
    Type '(props: ThemeProps) => SerializedStyles' is not assignable to type 'Interpolation<Theme>'.
      Type '(props: ThemeProps) => SerializedStyles' is not assignable to type 'FunctionInterpolation<Theme>'.
        Types of parameters 'props' and 'props' are incompatible.
          Property 'theme' is missing in type 'Theme' but required in type 'ThemeProps'.
  Overload 2 of 2, '(props: DefaultComponentProps<LinkTypeMap<{}, "a">>): Element', gave the following error.
    Type '(props: ThemeProps) => SerializedStyles' is not assignable to type 'Interpolation<Theme>'.  TS2769

    178 |                   underline="none"
    179 |                   onClick={toggleTawkHandler}
  > 180 |                   css={listLinkStyle}
        |                       ^
    181 |                 >
    182 |                   Contact Us
    183 |                 </ExternalLink>
```

**Comment by rickstaa** [comment]
@Andarist I did some further testing and the following syntax seems to be working after all:

```ts
const ButtonStyle = (theme: Theme) => css`
  background: ${theme.colors.primary};
`;
```

The working solution can be found in the [following sandbox](https://codesandbox.io/s/emotion-css-theme-i10gh). I think something else goes wrong while I'm defining my theme (I use material-ui + emotion.sh) I will investigate the problem myself and otherwise open a new issue.

**Comment by rickstaa** [comment]
An example of @Andarist   answer implemented in typescript can be found [here](https://codesandbox.io/s/emotion-css-theme-i10gh?file=/src/App.tsx:0-961).

**Comment by Andarist** [comment]
I think the answer has been given - if you have any further questions please open a discussion thread.

**Comment by StevenVerbiest** [comment]
@rickstaa Your link does not seems to be working. Can anyone give me an example of a typescript implementation? 
When I do 
```
const mixin = theme => css`
  color: ${theme.colors.primary}
`
```

I get the error `Parameter 'theme' implicitly has an 'any' type.`

And when I do 
```
const mixin = ({ theme }) => css`
  color: ${theme.colors.primary}
`
```

I get the error `Binding element 'theme' implicitly has an 'any' type.`

I've made a codesandbox of the problem:
https://codesandbox.io/s/emotion-11-css-api-typescript-6dv8r?file=/src/Div.styled.ts

**Comment by rickstaa** [comment]
@StevenVerbiest, I think I changed the code-sandbox name somewhere after I posted the example. I updated the link to https://codesandbox.io/s/emotion-material-ui-example-ni92b?file=/src/App.tsx. Let me know if that solves your problem.

**Comment by StevenVerbiest** [comment]
@rickstaa Thank you for the update! 

Your solution is similar to a workaround we had to do when implementing a custom theme interface (which is fixed in version 11). 

It would be nice if the `css` api would also use the custom Theme interface instead of having to pass it every time we use it.

**Comment by rickstaa** [comment]
@StevenVerbiest I'm not sure if I understand the improvement your suggesting. Maybe you can elaborate a bit on it. When you use solution 3 in the sandbox the CSS API is able to use the custom theme without us having to pass it. The type declaration is there just to make typescript happy. Due to how typescript is typed I think there is no way to get rid of this type of definition. 

I think the improvement your looking for is similar to what I asked in https://github.com/emotion-js/emotion/discussions/2291. After a discussion with @Andarist, I however agree that the module declaration overload is currently the cleanest solution we can support.

**Comment by linxianxi** [comment]
> @StevenVerbiest I'm not sure if I understand the improvement your suggesting. Maybe you can elaborate a bit on it. When you use solution 3 in the sandbox the CSS API is able to use the custom theme without us having to pass it. The type declaration is there just to make typescript happy. Due to how typescript is typed I think there is no way to get rid of this type of definition.
> 
> I think the improvement your looking for is similar to what I asked in #2291. After a discussion with @Andarist, I however agree that the module declaration overload is currently the cleanest solution we can support.

It would be nice if solution 2 didn't have to write the type every time. Style-component library can be done.

**Comment by gubi995** [comment]
I came across the same "problem" so I ended up doing this:

```tsx
const color = (color: ColorName) => ({ theme }: { theme: Theme }) =>
    css`
      ${theme.colors[color]}
    `;
    
const Button = styled.button`
     color: ${color('red')}
`    
```

**Comment by gusrodriguez** [comment]
You can also call the _useTheme_ hook to make available the theme object.

`
import { useTheme } from "@mui/material/styles"
const theme = useTheme();
...
const styles = {
    container: css`
      myStyle: ${theme.palette.primary.main};
  }
`

**Comment by Diableros** [comment]
> You can also call the _useTheme_ hook to make available the theme object.
> 
> ```
> import { useTheme } from "@mui/material/styles"
> const theme = useTheme();
> const styles = {
>     button: css`
>       background: ${theme.palette.primary.main};
>   }
> ```

Hooks don't work on top level (upper the component), or doesn't? How can use this approach on mixines upper the styled components?
